### DISCUSSION TITLE
Ghostty ZSH integration changes ZSH options, breaks aliases

### DISCUSSION THREAD

**Discussion by lykahb** [ghostty-org/ghostty#2847] [Q&A] [answered]
I followed the documentation for running nested shells (e.g, with `nix develop`) and inserted the integration snippet into `.zshrc`. As a result, my aliases stopped working. I set debugging output and saw that it changes quite a few options.

```
echo Before ghostty init
setopt

if [ -n "${GHOSTTY_RESOURCES_DIR}" ]; then
    builtin source "${GHOSTTY_RESOURCES_DIR}/shell-integration/zsh/ghostty-integration"
fi

echo After ghostty init
setopt
```

The output shows the options that changed

```
Before ghostty init
combiningchars
interactive
login
monitor
zle

After ghostty init
noaliases
combiningchars
interactive
localoptions
localpatterns
localtraps
login
monitor
zle
```

**Comment by mitchellh** [comment]
I've seen this before from the community and I believe there is some guideline to where the integration should be injected. I turned this to Q&A to get the answer. Once we figure it out, I'll update the docs to make it clearer.

If I'm mistaken and there is some other issue at play here, I can convert that into a more targeted issue, but some research is required by a zsh user to know more (I don't use or know zsh, unfortunately, and rely on the community to help maintain these shell integrations). 😄

**Reply by lykahb** [reply]
I believe that ghostty injects and executes the integration before the .zshrc is executed. This preserves the options. But when the integration is called from .zshrc, it updates the options. I moved the call around .zshrc and the problem persisted.

The explicit call is tested less. This looks like bug in the integration.

**Reply by lykahb** [reply]
The top-level call outside of a function defeats the purpose of the `-L` local scope on [this line](https://github.com/ghostty-org/ghostty/blob/main/src/shell-integration/zsh/ghostty-integration#L38). The `-L` limits the scope of changes for a function but lets it spill outside of the script. I can open a PR to put everything in the integration inside of a function.

I used this script to explore how the options persist in the scope
```
_func() {
  builtin emulate -L zsh -o no_warn_create_global -o no_aliases
  echo "Inside _func after updating options"
  _print_setopt  # Prints updated options
}

_print_setopt() {
  # The setopt is inside of a function in case that affects the scope
  setopt
}

echo "Before _func"
_print_setopt  # Prints original options

_func

echo "After_func"
_print_setopt  # Prints original options again
```

**Comment by gldtn** [comment]
I just got invited yesterday and while setting up Ghostty I ran into this as well. Functions were working, but all my aliases were not. I thought maybe I was doing something wrong. I'll report back if I discover anything that may be useful.

**Comment by shadow-absorber** [comment]
yeah have this problem too and had fun on the discord explaining it there.... should maybe have tried and find it here earlier

**Comment by shadow-absorber** [comment]
a workaround for tmux users is to launch tmux in another terminal emulator then detach and reattach in ghostty

why that works shadow is unsure of

**Comment by lykahb** [comment, marked as answer]
Opened a PR https://github.com/ghostty-org/ghostty/pull/2950
